### DISCUSSION TITLE
How to get the audio source of the virtual sound card output on Windows

### DISCUSSION THREAD

**Discussion by dycfan** [savonet/liquidsoap#2733] [User Support]
My liquidsoap runs on Windows, and I have a virtual sound card called "virtual-audio-capturer" on Windows that works. I want to use a script on Windows to get the audio source on this virtual sound card and send this audio source to icecast using a script on Windows. My question is: how to get the audio source of the "virtual-audio-capturer" virtual sound card using scripts on Windows.

Thank you for your urgent help!

**Comment by toots** [comment]
Hi @dycfan. Liquidsoap on windows uses portaudio for audio input.

I looked at the code and realized that, up to now, the user was not able to select the portaudio input or output device.

I have fixed this in the `rolling-release-v2.1.x` (as well as `main`).

If you download the windows build that is being prepare here: https://github.com/savonet/liquidsoap/actions/runs/3448157757, you should be able to:
* Call `liquidsoap.exe --list-portaudio-devices` to get the list of recognized devices
* Use `input.portaudio(device_id=xxx, ...)` where `device_id` is the ID of the device that you wish to use.

**Reply by dycfan** [reply]
Thank you very much for your guidance and help!

Following your guidance, I've used Liquidsoap on Windows 11 to successfully stream audio from my virtual sound card to my icecast server.

Thanks again!

Also, according to the official Liquidsoap documentation, I tried to call ffplay with input.external.rawaudi (I have the latest version of ffmpeg installed on Windows 11) to open the virtual sound card on my Windows, and my script code is as follows:

settings.log.stdout.set(true)
settings.log.file.set(true)
settings.log.file.path.set("C:/KuGou/test1.log")
devices="CABLEOutput (VB-Audio Virtual Cable)"
s = input.external.rawaudio("ffplay -f dshow audio="#{devices}" ")

output.icecast(%ffmpeg(format="mp3", %audio(codec="libmp3lame", b="320k")),
  host="192.168.5.6",port=8008,password="myhome",
  fallible=true,mount="myradio",s)

However, although the above script ran without errors, and the virtual sound card seems to have been successfully opened, unfortunately, the sound in the virtual sound card has not been successfully transcoded and streamed. Help: Is my method working? If the method works, how should the correct code be written?

Thank you! And please reply.
